solr/core/src/java/org/apache/solr/cluster/events/impl/ClusterEventProducerFactory.java

```diff
@@ -0,0 +1,173 @@
+    // install plugin registry listener
+    ContainerPluginsRegistry.PluginRegistryListener pluginListener = new ContainerPluginsRegistry.PluginRegistryListener() {
+      @Override
+      public void added(ContainerPluginsRegistry.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          ClusterEventListener listener = (ClusterEventListener) instance;
+          clusterEventProducer.registerListener(listener);
+        } else if (instance instanceof ClusterEventProducer) {
+          // replace the existing impl
+          if (cc.getClusterEventProducer() instanceof DelegatingClusterEventProducer) {
+            ((DelegatingClusterEventProducer) cc.getClusterEventProducer())
+                .setDelegate((ClusterEventProducer) instance);
+          } else {
+            log.warn("Can't configure plugin-based ClusterEventProducer while CoreContainer is still loading - " +
+                " using existing implementation {}", cc.getClusterEventProducer().getClass().getName());
+          }
+        }
+      }
+
+      @Override
+      public void deleted(ContainerPluginsRegistry.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          ClusterEventListener listener = (ClusterEventListener) instance;
+          clusterEventProducer.unregisterListener(listener);
+        } else if (instance instanceof ClusterEventProducer) {
+          // replace the existing impl with NoOp
+          if (cc.getClusterEventProducer() instanceof DelegatingClusterEventProducer) {
+            ((DelegatingClusterEventProducer) cc.getClusterEventProducer())
+                .setDelegate(new NoOpProducer(cc));
+          } else {
+            log.warn("Can't configure plugin-based ClusterEventProducer while CoreContainer is still loading - " +
+                " using existing implementation {}", cc.getClusterEventProducer().getClass().getName());
+          }
+        }
+      }
+
+      @Override
+      public void modified(ContainerPluginsRegistry.ApiInfo old, ContainerPluginsRegistry.ApiInfo replacement) {
+        added(replacement);
+        deleted(old);
+      }
+    };
+    plugins.registerListener(pluginListener);
+    created = true;
+    return clusterEventProducer;
+  }
+
+  private void transferListeners(ClusterEventProducer target, ContainerPluginsRegistry plugins) {
+    // stop capturing listener plugins
+    plugins.unregisterListener(initialPluginListener);
+    // transfer listeners that are already registered
+    listeners.forEach((type, listenersSet) -> {
+      listenersSet.forEach(listener -> target.registerListener(listener, type));
+    });
+    listeners.clear();
```

Comment:
Can't this be causing issues with methods iterating over `listeners` in the base class? Maybe need to synchronize `transferListeners`?